File: packages/gbnf-experiment/data/20260909T004354Z_a47a3057/ported_implementation/src/grammar-graph/grammar-graph-types.ts
import { validateNonEmpty } from '../utils/validate-non-empty.js';
import type { GraphNode } from './graph-node.js';
import type { GraphPointer } from './graph-pointer.js';
import type { Pointers } from './pointers.js';
import { RuleRef } from './rule-ref.js';

export const RuleType = {
  CHAR: 'char',
  CHAR_EXCLUDE: 'char_exclude',
  RULE_REF: 'rule_ref',
  END: 'end',
} as const;
export type RuleType = (typeof RuleType)[keyof typeof RuleType];

export type Range = [number, number];

export interface PrintOpts {
  pointers?: Pointers;
  colorize: (value: string | number, color: string) => string;
  showPosition?: boolean;
}

abstract class RuleWithListOfIntsOrRanges {
  value: (number | Range)[];

  constructor(value: (number | Range)[]) {
    // the incoming list is copied so later mutations of the rule that produced
    // it cannot leak into this one.
    this.value = [...value];
  }
}

export class RuleChar extends RuleWithListOfIntsOrRanges {
  type = RuleType.CHAR;
}

export class RuleCharExclude extends RuleWithListOfIntsOrRanges {
  type = RuleType.CHAR_EXCLUDE;
}

export class RuleEnd {
  type = RuleType.END;
}

export { RuleRef, validateNonEmpty };

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

// RuleRefs should never be exposed to the end user.
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

// ValidInput can either be a string, or a number indicating a code point.
// It CANNOT be a number representing a number; a number intended as input (like "8")
// should be passed in as a string.
export type ValidInput = string | number | number[];

export type ResolvedGraphPointer = GraphPointer<ResolvedRule>;
export type RootNode = Map<number, GraphNode>;

File: proposed-projects/round-trip-experiment/reverse/20260909T150124Z_a7cd2ee9/20260910T084451Z_b3565ede/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
from typing import Union

from ..utils.validate_non_empty import validate_non_empty

# A range of code points, inclusive on both ends: [start, end].
Range = list

# ValidInput can either be a string, or a number indicating a code point.
# It CANNOT be a number representing a number; a number intended as input (like "8")
# should be passed in as a string.
ValidInput = Union[str, int, list]


class RuleType:
    CHAR = "char"
    CHAR_EXCLUDE = "char_exclude"
    END = "end"
    REF = "ref"


class RuleEnd:
    def __init__(self) -> None:
        self.type = RuleType.END

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RuleEnd):
            return NotImplemented
        return True

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return "RuleEnd()"


class RuleChar:
    def __init__(self, value: list) -> None:
        self.type = RuleType.CHAR
        self.value = list(validate_non_empty(value))

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RuleChar):
            return NotImplemented
        return self.value == other.value

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return f"RuleChar(value={self.value!r})"


class RuleCharExclude:
    def __init__(self, value: list) -> None:
        self.type = RuleType.CHAR_EXCLUDE
        self.value = list(validate_non_empty(value))

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RuleCharExclude):
            return NotImplemented
        return self.value == other.value

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return f"RuleCharExclude(value={self.value!r})"
